File: src/ios/GIZPlugin.m
#import "GIZPlugin.h"

@implementation GIZPlugin

- (void)pluginInitialize {
    [super pluginInitialize];
    mBindDevice = nil;
    discoverWifiCommand = nil;
    discoverDeviceCommand = nil;
    scanDeviceList = @[];
}

// ----------------------------------- cordova 方法 ----------------------------------
- (void)init:(CDVInvokedUrlCommand*)command
{
    // 设置 SDK 委托
    [GizWifiSDK sharedInstance].delegate = self;
    
    // 设置 AppInfo
    NSDictionary* appInfo = @{@"appId": @"5fa1e143f0ce4c5cba05b682e833fe71", @"appSecret": @"d02788561db64b14a495f9a76f0d9e74"};
    
    // 设置要过滤的设备 productKey 列表。不过滤则直接传 nil
    NSArray *productInfo = [NSArray arrayWithObjects: @{@"productKey": @"f38e24d49d0e47a7934746171fdbf382",
                                                        @"productSecret": @"108c99829030439d876ac65ca148eb3e"}, nil];
    
    [GizWifiSDK startWithAppInfo:appInfo productInfo:productInfo cloudServiceInfo:nil autoSetDeviceDomain:NO];
    discoverWifiCommand = command;
}

- (void)userLogin:(CDVInvokedUrlCommand*)command
{
    NSString *userName = [command.arguments objectAtIndex:0];
    NSString *password = [command.arguments objectAtIndex:1];
    [[GizWifiSDK sharedInstance] userLogin:userName password:password];
    [self callback:command response:nil];
}

// 用户注册
- (void)userRegister:(CDVInvokedUrlCommand*)command
{
    NSString *username = [command.arguments objectAtIndex:0];
    NSString *password = [command.arguments objectAtIndex:1];
    NSString *code = [command.arguments objectAtIndex:2];
    NSString *type = [command.arguments objectAtIndex:3];
    GizUserAccountType accountType = GizUserNormal;
    if ([type isEqualToString:@"email"]){
        accountType = GizUserEmail;
    }
    if ([type isEqualToString:@"phone"]){
        accountType = GizUserPhone;
    }
    if ([type isEqualToString:@"normal"]){
        accountType = GizUserNormal;
    }
    [[GizWifiSDK sharedInstance] registerUser:username password:password verifyCode:code accountType:accountType];
    [self callback:command response:nil];
}

// 获取手机验证码
- (void)getSMSCode:(CDVInvokedUrlCommand*)command
{
    NSString *phone = [command.arguments objectAtIndex:0];
    [[GizWifiSDK sharedInstance] requestSendPhoneSMSCode:@"108c99829030439d876ac65ca148eb3e" phone:phone];
    [self callback:command response:nil];
}

// 匿名登录
- (void)userLoginAnonymous:(CDVInvokedUrlCommand*)command
{
    [[GizWifiSDK sharedInstance] userLoginAnonymous];
    [self callback:command response:nil];
}

// 获取绑定的设备
- (void)getBindDeviceList:(CDVInvokedUrlCommand*)command
{
    NSString *uid = [command.arguments objectAtIndex:0];
    NSString *token = [command.arguments objectAtIndex:1];
    [[GizWifiSDK sharedInstance] getBoundDevices:uid token:token];
    [self callback:command response:nil];
}

// 获取监听设备状态
- (void)getNotifyDeviceStatus:(CDVInvokedUrlCommand*)command
{
    NSDictionary *device = @{@"macAddress":mBindDevice.macAddress,@"did":mBindDevice.did,@"productKey":mBindDevice.productKey,@"ipAddress":mBindDevice.ipAddress,@"isLAN":@(mBindDevice.isLAN),@"netStatus":@(mBindDevice.netStatus),@"isBind":@(mBindDevice.isBind),@"productType":@(mBindDevice.productType),@"productName":mBindDevice.productName,@"remark":mBindDevice.remark,@"alias":mBindDevice.alias,@"isSubscribe":@(mBindDevice.isSubscribed)};
    [self callback:command response:device];
}

// 设置订阅
- (void)setSubscribe:(CDVInvokedUrlCommand*)command
{
    // NSString *isSubscribe = [command.arguments objectAtIndex:0];
    [mBindDevice setSubscribe:@"108c99829030439d876ac65ca148eb3e" subscribed:YES];
    [self callback:command response:nil];
}

// 发送控制命令
- (void) setCommand:(CDVInvokedUrlCommand*)command
{
    NSString *commandString = [command.arguments objectAtIndex:0];
    NSString *snString = [command.arguments objectAtIndex:1];
    int sn = [snString intValue];
    NSString *ResultString = [self jsonToHex: commandString];
    NSDictionary *data = @{@"binary": [self stringToByte: ResultString]};
    [mBindDevice write: data withSN:sn];
    [self callback:command response:nil];
}

// 解除绑定设备
- (void) unbindDevice:(CDVInvokedUrlCommand*)command
{
    NSString *uid = [command.arguments objectAtIndex:0];
    NSString *token = [command.arguments objectAtIndex:1];
    NSString *did = [command.arguments objectAtIndex:2];
    
    [[GizWifiSDK sharedInstance] unbindDevice:uid token:token did:did];
    [self callback:command response:nil];
}

// 绑定设备
-(void) bindDevice:(CDVInvokedUrlCommand*)command
{
    NSString *uid = [command.arguments objectAtIndex:0];
    NSString *token = [command.arguments objectAtIndex:1];
    NSString *mac = [command.arguments objectAtIndex:2];
    [[GizWifiSDK sharedInstance] bindRemoteDevice:uid token:token mac:mac productKey:@"f38e24d49d0e47a7934746171fdbf382" productSecret:@"108c99829030439d876ac65ca148eb3e" beOwner:YES];
    [self callback:command response:nil];
}

// 开启设备监听
-(void) startDeviceNotify:(CDVInvokedUrlCommand*)command
{
    if (mBindDevice != nil) {
        mBindDevice = nil;
    }
    
    NSString *mac = [command.arguments objectAtIndex:0];
    for (GizWifiDevice* dev in scanDeviceList) {
        if ([mac isEqualToString:dev.macAddress]) {
            mBindDevice = dev;
        }
    }
    
    discoverDeviceCommand = command;
    
    mBindDevice.delegate = self;
}

// ----------------------------------- cordova 方法 ----------------------------------

// ----------------------------------- 数据回调监听 ------------------------------------
// 实现系统事件通知回调
- (void)wifiSDK:(GizWifiSDK *)wifiSDK didNotifyEvent:(GizEventType)eventType eventSource:(id)eventSource eventID:(GizWifiErrorCode)eventID eventMessage: (NSString *)eventMessage {
    if(eventType == GizEventSDK) {
        // SDK发生异常的通知
        NSLog(@"SDK event happened: [%@] = %@", @(eventID), eventMessage);
    } else if(eventType == GizEventDevice) {
        // 设备连接断开时可能产生的通知
        GizWifiDevice* mDevice = (GizWifiDevice*)eventSource;
        NSLog(@"device mac %@ disconnect caused by %@", mDevice.macAddress, eventMessage);
    } else if(eventType == GizEventM2MService) {
        // M2M服务返回的异常通知
        NSLog(@"M2M domain %@ exception happened: [%@] = %@", (NSString*)eventSource, @(eventID), eventMessage);
    } else if(eventType == GizEventToken) {
        // token失效通知
        NSLog(@"token %@ expired: %@", (NSString*)eventSource, eventMessage);
    }
    NSLog(@"------------------------------------ init success -----------------------------------");
    NSDictionary *response = @{@"type":@"system",@"gizEventType":@(eventType)};
    [self keepCallback:discoverWifiCommand response:response];
}

// 登录回调
- (void)wifiSDK:(GizWifiSDK *)wifiSDK didUserLogin:(NSError *)result uid:(NSString *)uid token:(NSString *)token {
    NSDictionary *response = @{@"type":@"userLogin",@"gizEventType":@(result.code),@"uid":uid,@"token":token};
    [self keepCallback:discoverWifiCommand response:response];
}

// 注册回调
- (void)wifiSDK:(GizWifiSDK *)wifiSDK didRegisterUser:(NSError *)result uid:(NSString *)uid token:(NSString *)token {
    NSDictionary *response = @{@"type":@"userRegister",@"gizEventType":@(result.code),@"uid":uid,@"token":token};
    [self keepCallback:discoverWifiCommand response:response];
}

// 获取手机验证码回调
- (void)wifiSDK:(GizWifiSDK *)wifiSDK didRequestSendPhoneSMSCode:(NSError *)result token:(NSString *)token {
    NSDictionary *response = @{@"type":@"getSMSCode",@"gizEventType":@(result.code),@"token":token};
    [self keepCallback:discoverWifiCommand response:response];
}

// 接收设备列表变化上报pingcan
- (void)wifiSDK:(GizWifiSDK *)wifiSDK didDiscovered:(NSError *)result deviceList:(NSArray *)deviceList {
    NSMutableArray *devices = [NSMutableArray array];
    scanDeviceList = deviceList;
    for (GizWifiDevice* dev in deviceList) {
        NSDictionary *device = @{@"macAddress":dev.macAddress,@"did":dev.did,@"productKey": dev.productKey,@"ipAddress":dev.ipAddress,@"isLAN":@(dev.isLAN),@"netStatus":@(dev.netStatus),@"isBind":@(dev.isBind),@"productType":@(dev.productType),@"productName":dev.productName,@"remark":dev.remark,@"alias":dev.alias,@"isSubscribe":@(dev.isSubscribed)};
        [devices addObject:device];
    }
    NSDictionary *response = @{@"type":@"getDeviceList",@"gizEventType":@(result.code),@"deviceList": devices};
    [self keepCallback:discoverWifiCommand response:response];
}

// 解除绑定设备回调
- (void)wifiSDK:(GizWifiSDK *)wifiSDK didUnbindDevice:(NSError *)result did:(NSString *)did {
    NSDictionary *response = @{@"type":@"unbindDevice",@"gizEventType":@(result.code)};
    [self keepCallback:discoverWifiCommand response:response];
}

// 绑定设备回调
- (void)wifiSDK:(GizWifiSDK *)wifiSDK didBindDevice:(NSError *)result did:(NSString *)did {
    NSDictionary *response = @{@"type":@"bindDevice",@"gizEventType":@(result.code)};
    [self keepCallback:discoverWifiCommand response:response];
}

// 订阅回调
- (void)device:(GizWifiDevice *)device didSetSubscribe:(NSError *)result isSubscribed:(BOOL)isSubscribed {
    NSDictionary *response = @{@"type":@"setSubscribe",@"gizEventType":@(result.code)};
    [self keepCallback:discoverDeviceCommand response:response];
}

// 数据回调
- (void)device:(GizWifiDevice *)device didReceiveData:(NSError *)result data:(NSDictionary *)data withSN:(NSNumber *)sn {
    if(result.code == GIZ_SDK_SUCCESS) {
        NSLog(@"write successfully");
        if (data != nil) {
            NSData *binary = [data valueForKey:@"binary"];
            Byte *bytes = (Byte *)[binary bytes];
            NSString *dataString = [self byteToHexString: bytes];
            NSDictionary *response = @{@"type":@"write",@"gizEventType":@(result.code),@"result":[self hexToJson:dataString]};
            [self keepCallback:discoverDeviceCommand response:response];
        }
    } else {
        // 执行失败
    }
}
// ----------------------------------- 数据回调监听 ------------------------------------



// ----------------------------------- 基础方法 ------------------------------------
// keep callback
-(void)keepCallback:(CDVInvokedUrlCommand *) command response:(NSDictionary *) response
{
    CDVPluginResult *pluginResult = nil;
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: response];
    [pluginResult setKeepCallbackAsBool:TRUE];
    //通过cordova框架中的callBackID回调至JS的回调函数上
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

// callback
-(void)callback:(CDVInvokedUrlCommand *) command response:(NSDictionary *) response
{
    CDVPluginResult *pluginResult = nil;
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary: response];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

// hex to json
- (NSData *)hexToJson:(NSString *)hexString
{
    if (!hexString || [hexString length] == 0) {
        return nil;
    }
    
    NSMutableData *hexData = [[NSMutableData alloc] initWithCapacity:8];
    NSRange range;
    if ([hexString length] %2 == 0) {
        range = NSMakeRange(0,2);
    } else {
        range = NSMakeRange(0,1);
    }
    for (NSInteger i = range.location; i < [hexString length]; i += 2) {
        unsigned int anInt;
        NSString *hexCharStr = [hexString substringWithRange:range];
        NSScanner *scanner = [[NSScanner alloc] initWithString:hexCharStr];
        
        [scanner scanHexInt:&anInt];
        NSData *entity = [[NSData alloc] initWithBytes:&anInt length:1];
        [hexData appendData:entity];
        
        range.location += range.length;
        range.length = 2;
    }
    
    return hexData;
}

// 16进制转byte
-(NSData*)stringToByte:(NSString*)string
{
    NSString *hexString=[[string uppercaseString] stringByReplacingOccurrencesOfString:@" " withString:@""];
    if ([hexString length]%2!=0) {
        return nil;
    }
    Byte tempbyt[1]={0};
    NSMutableData* bytes=[NSMutableData data];
    for(int i=0;i<[hexString length];i++)
    {
        unichar hex_char1 = [hexString characterAtIndex:i]; ////两位16进制数中的第一位(高位*16)
        int int_ch1;
        if(hex_char1 >= '0' && hex_char1 <='9')
            int_ch1 = (hex_char1-48)*16;   //// 0 的Ascll - 48
        else if(hex_char1 >= 'A' && hex_char1 <='F')
            int_ch1 = (hex_char1-55)*16; //// A 的Ascll - 65
        else
            return nil;
        i++;
        
        unichar hex_char2 = [hexString characterAtIndex:i]; ///两位16进制数中的第二位(低位)
        int int_ch2;
        if(hex_char2 >= '0' && hex_char2 <='9')
            int_ch2 = (hex_char2-48); //// 0 的Ascll - 48
        else if(hex_char2 >= 'A' && hex_char2 <='F')
            int_ch2 = hex_char2-55; //// A 的Ascll - 65
        else
            return nil;
        
        tempbyt[0] = int_ch1+int_ch2;  ///将转化后的数放入Byte数组里
        [bytes appendBytes:tempbyt length:1];
    }
    return bytes;
}

// 字符串转16进制
-(NSString *)jsonToHex:(NSString *)string
{
    NSData *myD = [string dataUsingEncoding:NSUTF8StringEncoding];
    Byte *bytes = (Byte *)[myD bytes];
    NSString *hexStr=@"";
    
    for(int i=0;i<[myD length];i++)
    {
        NSString *newHexStr = [NSString stringWithFormat:@"%x",bytes[i]&0xff];///16进制数
        if([newHexStr length]==1)
            hexStr = [NSString stringWithFormat:@"%@0%@",hexStr,newHexStr];
        else
            hexStr = [NSString stringWithFormat:@"%@%@",hexStr,newHexStr];
    }
    NSString *headerString = @"dccd0202";
    long hexStringInt = [hexStr length]/2;
    NSString *hexStringLen = [self ToHex:hexStringInt];
    for (int i = 8; i > [hexStringLen length]; i--) {
        headerString = [NSString stringWithFormat:@"%@%@", headerString, @"0" ];
    }
    if([hexStringLen length]==1 || [hexStringLen length]==3 || [hexStringLen length]==5 || [hexStringLen length]==7) {
        hexStringLen = [NSString stringWithFormat:@"%@%@",@"0", hexStringLen];
    }
    headerString = [NSString stringWithFormat:@"%@%@",headerString, hexStringLen];
    hexStr = [NSString stringWithFormat:@"%@%@",headerString, hexStr];
    return hexStr;
}

// 10进制转16进制
- (NSString *)ToHex:(long)tmpid
{
    NSString *nLetterValue;
    NSString *str = @"";
    int ttmpig;
    for (int i =0; i<9; i++) {
        ttmpig=tmpid%16;
        tmpid=tmpid/16;
        switch (ttmpig)
        {
            case 10:
                nLetterValue =@"A";break;
            case 11:
                nLetterValue =@"B";break;
            case 12:
                nLetterValue =@"C";break;
            case 13:
                nLetterValue =@"D";break;
            case 14:
                nLetterValue =@"E";break;
            case 15:
                nLetterValue =@"F";break;
            default:nLetterValue=[[NSString alloc]initWithFormat:@"%i",ttmpig];
                
        }
        str = [nLetterValue stringByAppendingString:str];
        if (tmpid == 0) {
            break;
        }
    }
    return str;
}

// byte to hex
-(NSString *)byteToHexString:(Byte *)bytes
{
    NSString *hexStr=@"";
    for(int i=0;i<sizeof(bytes);i++)
    {
        NSString *newHexStr = [NSString stringWithFormat:@"%x",bytes[i]&0xff];///16进制数

        if([newHexStr length]==1)

            hexStr = [NSString stringWithFormat:@"%@0%@",hexStr,newHexStr];

        else

            hexStr = [NSString stringWithFormat:@"%@%@",hexStr,newHexStr];
    }
    return hexStr;
}
@end

File: src/android/GIZPlugin.java
package com.jieweifu.plugins.giz;

import android.content.ComponentName;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.util.Log;

import com.gizwits.gizwifisdk.api.GizWifiDevice;
import com.gizwits.gizwifisdk.api.GizWifiSDK;
import com.gizwits.gizwifisdk.enumration.GizEventType;
import com.gizwits.gizwifisdk.enumration.GizUserAccountType;
import com.gizwits.gizwifisdk.enumration.GizWifiErrorCode;
import com.gizwits.gizwifisdk.listener.GizWifiDeviceListener;
import com.gizwits.gizwifisdk.listener.GizWifiSDKListener;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaArgs;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class GIZPlugin extends CordovaPlugin {
    private Context mContext;

    private static GIZPlugin instance;

    public GIZPlugin() {
        instance = this;
    }

    private CallbackContext discoverWifiCallback;

    private CallbackContext discoverDeviceCallback;

    private static GizWifiDevice mBindDevice;

    private List<GizWifiDevice> scanDeviceList;

    private String appId = "";

    private String appSecret = "";

    private String productKey = "";

    private String productSecret = "";

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        mContext = cordova.getActivity().getApplicationContext();
    }

    @Override
    public boolean execute(final String action, final CordovaArgs args, final CallbackContext callbackContext) throws JSONException {
        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Method method = GIZPlugin.class.getDeclaredMethod(action, CordovaArgs.class, CallbackContext.class);
                    method.invoke(GIZPlugin.this, args, callbackContext);
                } catch (Exception e) {
                    Log.e("", e.toString());
                }
            }
        });
        return true;
    }

    void init(CordovaArgs args, CallbackContext callbackContext) throws JSONException, PackageManager.NameNotFoundException {
        discoverWifiCallback = callbackContext;

        ApplicationInfo info = mContext.getPackageManager().getApplicationInfo(mContext.getPackageName(), PackageManager.GET_META_DATA);

        appId = info.metaData.getString("APP_ID") ;
        appSecret = info.metaData.getString("APP_SECRET");
        productKey = info.metaData.getString("PRODUCT_KEY");
        productSecret = info.metaData.getString("PRODUCT_SECRET");
        GizWifiSDK.sharedInstance().setListener(wifiListener);
        ConcurrentHashMap<String, String> appInfo = new ConcurrentHashMap();
        appInfo.put("appId", appId);
        appInfo.put("appSecret", appSecret);

        List<ConcurrentHashMap<String, String>> productInfo = new ArrayList();
        ConcurrentHashMap<String, String> product = new ConcurrentHashMap();
        product.put("productKey", productKey);
        product.put("productSecret", productSecret);
        productInfo.add(product);
        GizWifiSDK.sharedInstance().startWithAppInfo(mContext, appInfo, productInfo, null, false);
    }

    // 正常登录
    void userLogin(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String username = args.getString(0);
        String password = args.getString(1);
        GizWifiSDK.sharedInstance().userLogin(username, password);
        callbackContext.success();
    }

    // 注册
    void userRegister(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String username = args.getString(0);
        String password = args.getString(1);
        String code = args.getString(2);
        String type = args.getString(3);
        GizUserAccountType accountType = GizUserAccountType.GizUserNormal;
        if (type.equals("email")) {
            accountType = GizUserAccountType.GizUserEmail;
        } else if (type.equals("phone")) {
            accountType = GizUserAccountType.GizUserPhone;
        } else if (type.equals("normal")) {
            accountType = GizUserAccountType.GizUserNormal;
        }
        GizWifiSDK.sharedInstance().registerUser(username, password, code, accountType);
        callbackContext.success();
    }

    // 获取手机验证码
    void getSMSCode(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String phone = args.getString(0);
        GizWifiSDK.sharedInstance().requestSendPhoneSMSCode (appSecret, phone);
        callbackContext.success();
    }

    // 匿名登录
    void userLoginAnonymous(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        GizWifiSDK.sharedInstance().userLoginAnonymous();
        callbackContext.success();
    }

    // 获取绑定的设备
    void getBindDeviceList(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String uid = args.getString(0);
        String token = args.getString(1);
        GizWifiSDK.sharedInstance().getBoundDevices(uid, token);
        callbackContext.success();
    }

    // 获取监听设备状态
    void getNotifyDeviceStatus(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        JSONObject deviceObj = new JSONObject();
        deviceObj.put("macAddress", mBindDevice.getMacAddress());
        deviceObj.put("did", mBindDevice.getDid());
        deviceObj.put("productKey", mBindDevice.getProductKey());
        deviceObj.put("ipAddress", mBindDevice.getIPAddress());
        deviceObj.put("isLAN", mBindDevice.isLAN());
        deviceObj.put("netStatus", mBindDevice.getNetStatus());
        deviceObj.put("isBind", mBindDevice.isBind());
        deviceObj.put("productType", mBindDevice.getProductType());
        deviceObj.put("productName", mBindDevice.getProductName());
        deviceObj.put("remark", mBindDevice.getRemark());
        deviceObj.put("alias", mBindDevice.getAlias());
        deviceObj.put("isSubscribe", mBindDevice.isSubscribed());
        callbackContext.success(deviceObj);
    }

    // 设置订阅
    void setSubscribe(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String isSubscribe = args.getString(0);
        mBindDevice.setSubscribe("",  Boolean.parseBoolean(isSubscribe));
        callbackContext.success();
    }

    // 发送控制命令
    void setCommand(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String commandString = args.getString(0);
        Integer sn = args.getInt(1);
        ConcurrentHashMap<String, Object> command = jsonStringToByte(commandString);
        mBindDevice.write(command, sn);
        callbackContext.success();
    }

    // 解除绑定设备
    void unbindDevice(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String uid = args.getString(0);
        String token = args.getString(1);
        String did = args.getString(2);
        GizWifiSDK.sharedInstance().unbindDevice (uid, token, did);
        callbackContext.success();
    }

    // 绑定设备
    void bindDevice(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String uid = args.getString(0);
        String token = args.getString(1);
        String mac = args.getString(2);
        GizWifiSDK.sharedInstance().bindRemoteDevice (uid, token, mac, productKey, productSecret);
        callbackContext.success();
    }

    // 开启设备监听
    void startDeviceNotify(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        discoverDeviceCallback = callbackContext;

        if (mBindDevice != null) {
            mBindDevice = null;
        }

        String mac = args.getString(0);
        for (GizWifiDevice device: scanDeviceList) {
            if (mac.equals(device.getMacAddress())) {
                mBindDevice = device;
                mBindDevice.setListener(deviceListener);
            }
        }
    }

    // WIFI Listener
    private GizWifiSDKListener wifiListener = new GizWifiSDKListener() {
        @Override
        public void didNotifyEvent(GizEventType eventType, Object eventSource, GizWifiErrorCode eventID, String eventMessage) {
            try {
                JSONObject response = new JSONObject();
                response.put("type", "system");
                response.put("gizEventType", eventType.name());
                keepCallbackResult(discoverWifiCallback, response);
            } catch (JSONException e) {

            }
        }

        // 登录回调
        @Override
        public void didUserLogin(GizWifiErrorCode result, String uid, String token) {
            try {
                JSONObject response = new JSONObject();
                response.put("type", "userLogin");
                response.put("gizEventType", result.name());
                response.put("uid", uid);
                response.put("token", token);
                keepCallbackResult(discoverWifiCallback, response);
            } catch (JSONException e) {

            }
        }

        // 注册回调
        @Override
        public void didRegisterUser(GizWifiErrorCode result, String uid, String token) {
            try {
                JSONObject response = new JSONObject();
                response.put("type", "userRegister");
                response.put("gizEventType", result.name());
                response.put("uid", uid);
                response.put("token", token);
                keepCallbackResult(discoverWifiCallback, response);
            } catch (JSONException e) {

            }
        }

        // 获取手机验证码回调
        @Override
        public void didRequestSendPhoneSMSCode(GizWifiErrorCode result, String token) {
            try {
                JSONObject response = new JSONObject();
                response.put("type", "getSMSCode");
                response.put("gizEventType", result.name());
                response.put("token", token);
                keepCallbackResult(discoverWifiCallback, response);
            } catch (JSONException e) {

            }
        }

        // 获取设备列表回调
        @Override
        public void didDiscovered(GizWifiErrorCode result, List<GizWifiDevice> deviceList) {
            try {
                JSONObject response = new JSONObject();
                response.put("type", "getDeviceList");
                response.put("gizEventType", result.name());

                JSONArray jsonArray = new JSONArray();

                for(GizWifiDevice device: deviceList) {
                    JSONObject deviceObj = new JSONObject();
                    deviceObj.put("macAddress", device.getMacAddress());
                    deviceObj.put("did", device.getDid());
                    deviceObj.put("productKey", device.getProductKey());
                    deviceObj.put("ipAddress", device.getIPAddress());
                    deviceObj.put("isLAN", device.isLAN());
                    deviceObj.put("netStatus", device.getNetStatus());
                    deviceObj.put("isBind", device.isBind());
                    deviceObj.put("productType", device.getProductType());
                    deviceObj.put("productName", device.getProductName());
                    deviceObj.put("remark", device.getRemark());
                    deviceObj.put("alias", device.getAlias());
                    deviceObj.put("isSubscribe", device.isSubscribed());
                    jsonArray.put(deviceObj);
                }

                response.put("deviceList", jsonArray);

                scanDeviceList = deviceList;
                keepCallbackResult(discoverWifiCallback, response);
            } catch (JSONException e) {

            }
        }

        // 解除绑定设备回调
        @Override
        public void didUnbindDevice(GizWifiErrorCode result, String did) {
            try {
                JSONObject response = new JSONObject();
                response.put("type", "unbindDevice");
                response.put("gizEventType", result.name());
                keepCallbackResult(discoverWifiCallback, response);
            } catch (JSONException e) {

            }
        }

        // 绑定设备回调
        @Override
        public void didBindDevice(GizWifiErrorCode result, String did) {
            try {
                JSONObject response = new JSONObject();
                response.put("type", "bindDevice");
                response.put("gizEventType", result.name());
                keepCallbackResult(discoverWifiCallback, response);
            } catch (JSONException e) {

            }
        }
    };

    // 设备监听Listener
    private GizWifiDeviceListener deviceListener = new GizWifiDeviceListener() {
        @Override
        public void didSetSubscribe(GizWifiErrorCode result, GizWifiDevice device, boolean isSubscribed) {
            try {
                JSONObject response = new JSONObject();
                response.put("type", "setSubscribe");
                response.put("gizEventType", result.name());
                keepCallbackResult(discoverDeviceCallback, response);
            } catch (JSONException e) {

            }
        }

        @Override
        public void didReceiveData(GizWifiErrorCode result, GizWifiDevice device, ConcurrentHashMap<String, Object> dataMap, int sn) {
            try {
                if (result == GizWifiErrorCode.GIZ_SDK_SUCCESS) {
                    // 查询成功
                    if (dataMap != null && dataMap.size() > 0) {
                        if (dataMap.toString().contains("binary")) {
                            byte[] byteStrx = (byte[])dataMap.get("binary");
                            JSONObject obj = byteToJson(byteStrx);
                            JSONObject response = new JSONObject();
                            response.put("type", "write");
                            response.put("gizEventType", result.name());
                            response.put("result", obj);
                            keepCallbackResult(discoverDeviceCallback, response);
                        }
                    }
                } else {
                    // 查询失败
                    Log.i("getDeviceStatus", "fail");
                }
            } catch (JSONException e) {

            }
        }
    };

    private void keepCallbackResult(CallbackContext callbackContext, JSONObject data) {
        PluginResult result = new PluginResult(PluginResult.Status.OK, data);
        result.setKeepCallback(true);
        callbackContext.sendPluginResult(result);
    }

    private static JSONObject byteToJson(byte[] data) throws JSONException {
        int len = data.length;
        int i = 8;
        String s = "";
        while(i < len) {
            int b = data[i] & 255;
            char c = (char) b;
            s = s + String.valueOf(c);
            i ++;
            if (i == len) {
                break;
            }
        }
        return new JSONObject(s);
    }

    private static ConcurrentHashMap<String, Object> jsonStringToByte(String commandString) throws JSONException {
        String str = "dccd0202";
        int strLen = commandString.length();
        String hexLenStr = Integer.toHexString(strLen);

        for (int i = 8; i > hexLenStr.length(); i--) {
            str = str + "0";
        }

        str = str + hexLenStr;

        for (int i = 0; i < strLen; i++) {
            int ch = (int) commandString.charAt(i);
            String s4 = Integer.toHexString(ch);
            str = str + s4;
        }

        byte[] bytes = new byte[str.length() / 2];
        for (int i = 0; i < str.length() / 2; i++) {
            String subStr = str.substring(i * 2, i * 2 + 2);
            bytes[i] = (byte) Integer.parseInt(subStr, 16);
        }

        ConcurrentHashMap<String, Object> command = new ConcurrentHashMap();
        command.put("binary", bytes);
        return command;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        instance = null;
    }
}